src/rulesets/Base/array.jl

```diff
@@ -37,6 +37,12 @@ function rrule(
     return Base.vect(X...), vect_pullback
 end
 
+# This rule isn't always correct, but Zygote and Diffractor fall over completely without it.
```

Comment:
Is it correct to say

```suggestion
# This rule isn't certain to be correct, if someone has overloaded `Base.vect` for their type
# in a way that has a non-identity derivative; but that is a very weird thing to do.
# so we are confident that this is safe.
#  Zygote and Diffractor fall over completely without this rule.
```

Reply:
I'm unsure about the correctness of this statement, because I don't fully understand the internals of `vect`. In particular, I don't have a good handle on when the objects being put inside the vector are modified and when they "pass straight through". Do you understand how this works in general?

Reply:
I thought it converted to the type from `mapreduce(typeof, Base.promote_type, ([1 2], [3,4]'))`, but not 100%.

I also think the comment should just say something like "Can't currently apply ProjectTo to arbitrary types, so leave them alone" as I think that's the extent of the non-correctness -- something might be promoted `vect` in a way that the rule won't un-do, but this is a pretty mild form of incorrectness. Unless you think it can go more seriously wrong?

Reply:
ok cool.
Yeah, i am fine with missing some `ProjectTo`s, especially in "nonmathematical" operations like this.
As long as all mathematical operations do the right thing when creating the cotangents then later things just moving them around don't have to worry.

I still approve this PR, but some good comment needs to be added.
You can't just say "isn't always correct" without explaining anything about when it isn't

Reply:
>  You can't just say "isn't always correct" without explaining anything about when it isn't

Haha that's a fair point. I'll improve the comment.

> As long as all mathematical operations do the right thing when creating the cotangents then later things just moving them around don't have to worry.

I'm not totally convinced this is true. The case that concerns me is somehow a sparse array getting promoted to a dense array in this operation, and the rule for whatever function which produced the sparse array receiving a dense (little-t) tangent matrix on the reverse pass, and not knowing what to do with it. I'm still not 100% up to speed with the new state of the world though, so possibly I'm missing a detail 🤷